### ISSUE TITLE
How to implement a reduction operation for NCCL AllReduce that is not associative or commutative

### ISSUE THREAD

**Issue by hamidralmasi** [NVIDIA/nccl#633]
I want to perform allreduce with a reduction operation that is not associative or commutative, say, median, and see its effect on the accuracy of training with Tensorflow or Pytorch. I have been looking into NCCL's ring implementation and I cannot see a way to modify it to do this because I assume for this case, all data should be present for reduction and this is not how the scatter-reduce steps in the ring algorithm work with data. I have a few questions about this and I would appreciate any help since I'm new to this:

1. Is there a way to perform such an operation in allreduce with NCCL?
2. In that case, what would be the options for the place performing the reduction? i.e. CPU/GPU in one of the ranks or all ranks, etc.
3. Are there any NCCL plugins that are doing something similar?

Thank you so much for your help!

**Comment by sjeaugey** [comment]
I think the easiest is to perform an allgather, then have each GPU process the whole data set to perform the desired operation in the desired order.

**Comment by stephenmsachs** [comment]
### Issue Cleanup: Helping Us Focus on Current Challenges

  We're reviewing older issues to ensure we prioritize the most relevant and active ones. Since this issue hasn't seen updates in over 6 months, we'll be closing it for now.

  ,*This change helps us focus our efforts on addressing any current issues our users are facing.* If this issue still affects you, please don't hesitate to reopen it with a quick update (e.g., "Still relevant on [version=X]").
  Thanks for your understanding and for contributing to NCCL.
